### ISSUE TITLE
[refactoring] Invalid thread access on editor.isDirty() 

### ISSUE THREAD

**Issue by dhuebner** [eclipse/xtext-eclipse#298]
Initializing rename refactoring and having plugin manifest editor opened cause an Invalid Thread Access exception

Xtext 2.12 
PDE 3.9.100

```
java.lang.reflect.InvocationTargetException
	at org.eclipse.jface.operation.ModalContext.run(ModalContext.java:398)
	at org.eclipse.jface.dialogs.ProgressMonitorDialog.run(ProgressMonitorDialog.java:481)
	at org.eclipse.ui.internal.progress.ProgressMonitorJobsDialog.run(ProgressMonitorJobsDialog.java:242)
	at org.eclipse.ui.internal.progress.ProgressManager$3.run(ProgressManager.java:895)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:70)
	at org.eclipse.ui.internal.progress.ProgressManager.busyCursorWhile(ProgressManager.java:930)
	at org.eclipse.ui.internal.progress.ProgressManager.busyCursorWhile(ProgressManager.java:905)
	at org.eclipse.ui.internal.progress.ProgressManager.run(ProgressManager.java:1078)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.totalSync(SyncUtil.java:59)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.totalSync(SyncUtil.java:52)
	at org.eclipse.xtext.ui.refactoring.ui.DefaultRenameElementHandler.execute(DefaultRenameElementHandler.java:70)
	at org.eclipse.ui.internal.handlers.HandlerProxy.execute(HandlerProxy.java:295)
	at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:90)
	at sun.reflect.GeneratedMethodAccessor160.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:55)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:282)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:264)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:132)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:152)
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:494)
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:488)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:210)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.executeCommand(KeyBindingDispatcher.java:286)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.press(KeyBindingDispatcher.java:507)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.processKeyEvent(KeyBindingDispatcher.java:558)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.filterKeySequenceBindings(KeyBindingDispatcher.java:378)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.access$0(KeyBindingDispatcher.java:324)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher$KeyDownFilter.handleEvent(KeyBindingDispatcher.java:86)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
	at org.eclipse.swt.widgets.Display.filterEvent(Display.java:1097)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4251)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1501)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1524)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1509)
	at org.eclipse.swt.widgets.Widget.sendKeyEvent(Widget.java:1538)
	at org.eclipse.swt.widgets.Widget.sendKeyEvent(Widget.java:1534)
	at org.eclipse.swt.widgets.Canvas.sendKeyEvent(Canvas.java:515)
	at org.eclipse.swt.widgets.Composite.keyDown(Composite.java:617)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:5735)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSendSuper(Native Method)
	at org.eclipse.swt.widgets.Widget.callSuper(Widget.java:227)
	at org.eclipse.swt.widgets.Widget.windowSendEvent(Widget.java:2140)
	at org.eclipse.swt.widgets.Shell.windowSendEvent(Shell.java:2402)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:5797)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSendSuper(Native Method)
	at org.eclipse.swt.widgets.Display.applicationSendEvent(Display.java:5226)
	at org.eclipse.swt.widgets.Display.applicationProc(Display.java:5375)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSend(Native Method)
	at org.eclipse.swt.internal.cocoa.NSApplication.sendEvent(NSApplication.java:128)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3703)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$4.run(PartRenderingEngine.java:1121)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1022)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:150)
	at org.eclipse.ui.internal.Workbench$5.run(Workbench.java:693)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:610)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:138)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:673)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:610)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1519)
Caused by: org.eclipse.swt.SWTException: Invalid thread access
	at org.eclipse.swt.SWT.error(SWT.java:4533)
	at org.eclipse.swt.SWT.error(SWT.java:4448)
	at org.eclipse.swt.SWT.error(SWT.java:4419)
	at org.eclipse.swt.widgets.Widget.error(Widget.java:799)
	at org.eclipse.swt.widgets.Widget.checkWidget(Widget.java:589)
	at org.eclipse.swt.widgets.Widget.getData(Widget.java:865)
	at org.eclipse.ui.forms.DetailsPart.getCurrentPage(DetailsPart.java:138)
	at org.eclipse.ui.forms.DetailsPart.isDirty(DetailsPart.java:163)
	at org.eclipse.ui.forms.ManagedForm.isDirty(ManagedForm.java:236)
	at org.eclipse.ui.forms.editor.FormPage.isDirty(FormPage.java:254)
	at org.eclipse.pde.internal.ui.editor.PDEFormEditor.computeDirtyState(PDEFormEditor.java:589)
	at org.eclipse.pde.internal.ui.editor.PDEFormEditor.isDirty(PDEFormEditor.java:583)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.reconcileAllEditors(SyncUtil.java:104)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.doReconcileAndBuild(SyncUtil.java:75)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.access$0(SyncUtil.java:71)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil$1.run(SyncUtil.java:62)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:119)
Root exception:
org.eclipse.swt.SWTException: Invalid thread access
	at org.eclipse.swt.SWT.error(SWT.java:4533)
	at org.eclipse.swt.SWT.error(SWT.java:4448)
	at org.eclipse.swt.SWT.error(SWT.java:4419)
	at org.eclipse.swt.widgets.Widget.error(Widget.java:799)
	at org.eclipse.swt.widgets.Widget.checkWidget(Widget.java:589)
	at org.eclipse.swt.widgets.Widget.getData(Widget.java:865)
	at org.eclipse.ui.forms.DetailsPart.getCurrentPage(DetailsPart.java:138)
	at org.eclipse.ui.forms.DetailsPart.isDirty(DetailsPart.java:163)
	at org.eclipse.ui.forms.ManagedForm.isDirty(ManagedForm.java:236)
	at org.eclipse.ui.forms.editor.FormPage.isDirty(FormPage.java:254)
	at org.eclipse.pde.internal.ui.editor.PDEFormEditor.computeDirtyState(PDEFormEditor.java:589)
	at org.eclipse.pde.internal.ui.editor.PDEFormEditor.isDirty(PDEFormEditor.java:583)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.reconcileAllEditors(SyncUtil.java:104)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.doReconcileAndBuild(SyncUtil.java:75)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil.access$0(SyncUtil.java:71)
	at org.eclipse.xtext.ui.refactoring.ui.SyncUtil$1.run(SyncUtil.java:62)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:119)

```

**Comment by cdietrich** [comment]
@dhuebner can you give steps what you are exactly doing. i cannot reproduce this on a first glance

**Comment by dhuebner** [comment]
@cdietrich 
I already wrote how to reproduce. You need to have an Plugin PDE Editor open and try to start rename refactoring in an Xtext Editor.
From the stacktrace you can see that:
totalSync() calls doReconcileAndBuild() in a separate thread 
doReconcileAndBuild()  SyncUtil Line: 104 calls editor.isDirty() 
DeatilsPart line: 138 calls control.getData() as from javadoc shuld be UI Thread: ` ERROR_THREAD_INVALID_ACCESS - when called from the wrong thread`
It's pretty clear to me a bug. And can be fixed by moving isDirty check in SyncUtil#reconcileAllEditors to  line 109, so it's executed in UI thread.

**Comment by cdietrich** [comment]
but i actually cant reproduce it

- create a plugin
- create a xtend file
- refactor the xtend file
works fine

**Comment by dhuebner** [comment]
@cdietrich Maybe writing a test would help?

**Comment by cdietrich** [comment]
i assume it has to do with the versions of pde and platform.
but i can reprodtuce this neither on oxygen nor on eclipse 3.x on my customer site
but i am not even sure how to get that PDEFormEditor
cause i only get the manifest editor offered

**Comment by cdietrich** [comment]
just could reproduce it

**Comment by dhuebner** [comment]
Good

**Comment by cdietrich** [comment]
fixed in 2.13
